File: graph/Leetcode-46-Permutations/Leetcode-46-Permutations.cpp
class Solution {
public:
    void backtrack(vector<vector<int>>& pers, vector<int>& per, vector<int>& nums){
        if(per.size() == nums.size()){
            pers.push_back(per);
        }else{
            for(int e : nums){
                if(find(per.begin(), per.end(), e) != per.end()){
                    continue;
                }
                per.push_back(e);
                backtrack(pers, per, nums);
                per.pop_back();
            }
        }
    };
    
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> pers;
        vector<int> per;
        
        backtrack(pers, per, nums);
        
        return pers;
    }
};

File: 2024상반기/2월1주차/leetcode-46-permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        answer = []
        def permutation(li):
            if len(li) == len(nums):
                answer.append(li[:])
                return
            for n in nums:
                if n not in li:
                    li.append(n)
                    permutation(li)
                    li.pop()
        permutation([])
        return answer